### PR TITLE
feat(shim-kvm): support old CPUs in KVM mode

### PR DESCRIPTION
**Pull request by haraldh**
Fixes: https://github.com/enarx/enarx/issues/1468

### CODE REVIEW

**Comment by npmccallum** [comment]
@haraldh The `rdrand` patch makes me nervous. Also, `rdrand` is **really** old. Do we really need that?

**Comment by haraldh** [comment]
we don't need it, but it does no real harm, because SNP is guarded

File: internal/shim-sev/src/gdt.rs [hunk 1]
```diff
@@ -170,5 +172,11 @@ pub unsafe fn init() {
...
     let base = VirtAddr::new(TSS.deref() as *const _ as u64);
     KernelGsBase::write(base);
-    GS::write_base(base);
+
+    if *CPU_HAS_FSGSBASE {
+        Cr4::update(|f| *f |= Cr4Flags::FSGSBASE);
+        GS::write_base(base);
+    } else {
+        GsBase::write(base);
+    }
```
Comment (bstrie) [inline internal/shim-sev/src/gdt.rs:181]:
```suggestion
    if *CPU_HAS_FSGSBASE {
        Cr4::update(|f| *f |= Cr4Flags::FSGSBASE);
    }
    GsBase::write(base);
```

**Review by npmccallum — APPROVED** [review]

File: internal/shim-sev/src/hostcall.rs [hunk 2]
```diff
@@ -251,28 +258,44 @@ impl Handler for HostCall<'_> {
         match code {
             syscall::ARCH_SET_FS => {
                 // FIXME: check `addr` value
-                unsafe {
-                    FS::write_base(VirtAddr::new(addr));
+                if *CPU_HAS_FSGSBASE {
+                    unsafe {
```
Comment (bstrie) [inline internal/shim-sev/src/hostcall.rs:262]: Is that FIXME the safety comment here?

File: internal/shim-sev/src/hostcall.rs [hunk 3]
```diff
@@ -251,28 +258,44 @@ impl Handler for HostCall<'_> {
...
+                    FsBase::read().as_u64()
+                };
                 Ok(())
             }
             syscall::ARCH_SET_GS => {
                 // FIXME: check `addr` value
-                unsafe {
-                    GS::write_base(VirtAddr::new(addr));
+                if *CPU_HAS_FSGSBASE {
+                    unsafe {
```
Comment (bstrie) [inline internal/shim-sev/src/hostcall.rs:283]: Same here.

File: internal/shim-sev/src/random.rs [hunk 4]
```diff
@@ -2,14 +2,31 @@
...
 /// Get a random number
 pub fn random() -> u64 {
     let mut r: u64 = 0;
 
-    for _ in 0..1024 {
-        if unsafe { core::arch::x86_64::_rdrand64_step(&mut r) } == 1 {
-            return r;
+    if *CPU_HAS_RDRAND {
+        for _ in 0..1024 {
+            if unsafe { core::arch::x86_64::_rdrand64_step(&mut r) } == 1 {
```
Comment (bstrie) [inline internal/shim-sev/src/random.rs:18]: Safe because we're ensuring that rdrand exists.

File: internal/shim-sev/src/random.rs [hunk 5]
```diff
@@ -2,14 +2,31 @@
...
+            }
         }
+    } else {
+        // This is an absolute fallback for old CPUs to be able to run in KVM simulation mode
+
+        if snp_active() {
+            panic!("No rdrand on SNP");
+        }
+
+        return unsafe { core::arch::x86_64::_rdtsc() };
```
Comment (bstrie) [inline internal/shim-sev/src/random.rs:29]: I'm not actually sure why this function is stdlib function is unsafe, can you think of a reason?

**Pull request merged**